cli/serve/serve.go

```diff
@@ -244,9 +287,41 @@ func serverMain(args []string, c cli.Config) error {
 		log.Info("Now listening on ", addr)
 		return http.ListenAndServe(addr, nil)
 	}
+	if conf.MutualTLSCAFile != "" {
+		clientPool, err := helpers.LoadPEMCertPool(conf.MutualTLSCAFile)
+		if err != nil {
+			return fmt.Errorf("failed to load mutual TLS CA file: %s", err)
+		}
 
-	log.Info("Now listening on https://", addr)
+		server := http.Server{
+			Addr: addr,
+			TLSConfig: &tls.Config{
+				ClientAuth: tls.RequireAndVerifyClientCert,
+				ClientCAs:  clientPool,
+			},
+		}
 
+		if conf.MutualTLSDNRegex != "" {
+			log.Debugf("Requiring DN matches regex %s for client connections", conf.MutualTLSDNRegex)
+			re := regexp.MustCompile(conf.MutualTLSDNRegex)
+			server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if r != nil && r.TLS != nil {
+					if len(r.TLS.PeerCertificates) > 0 {
```

Comment:
This can all be one check: `if r != nil && r.TLS != nil && len(r.TLS.PeerCertificates) > 0`